Add support for Blackrock NSx files with per-sample PTP nanosecond timestamps

cboulay opened the pull request:
Fixes #1332

Blackrock's newest signal acquisition devices use nanosecond timestamps from clocks synchronized with PTP. This allows synchronization between devices that might have slightly different clock rates and not capture exactly 30_000 samples per reference second.

Individual per-sample timestamps are stored in a data header, per sample.

Previously, the presence of a header in the data block would indicate the start of a new segment, and methods would look for the header to start a new memmap. That previous solution, when applied to millions of headers in the PTP-enabled files, led to OS-level problems by creating too many memmaps.

This PR allows the new PTP-enabled NSx files to be loaded.

Note: the time range of data files no longer start at 0! They start at whatever the monotonic clock was when the NSx file was recorded. We can't simply re-zero the timestamps per file because there might be multiple sets of files that were recorded simultaneously (i.e., from different pieces of hardware) but didn't start at the exact same time. We could re-zero with the common minimum across all devices, but the BlackrockRawIO loader doesn't support multi-device ns6 files. For now, it's best to leave it to the user to load the K files manually then find their own common t-zero among those K files.

pep8speaks commented:
Hello @cboulay! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`neo/rawio/blackrockrawio.py`](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py):

> [Line 324:1](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L324): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 326:17](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L326): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 381:21](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L381): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 458:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L458): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 966:1](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L966): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 983:17](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L983): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 987:13](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L987): [E121](https://duckduckgo.com/?q=pep8%20E121) continuation line under-indented for hanging indent
> [Line 1002:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1002): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 99 characters)
> [Line 1004:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1004): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 99 characters)
> [Line 1007:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1007): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 99 characters)
> [Line 1014:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1014): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 99 characters)
> [Line 1057:1](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1057): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1067:17](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1067): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 1071:13](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1071): [E121](https://duckduckgo.com/?q=pep8%20E121) continuation line under-indented for hanging indent
> [Line 1082:100](https://github.com/NeuralEnsemble/python-neo/blob/145f7579a36e178b54a866460a021a13ecf02c82/neo/rawio/blackrockrawio.py#L1082): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 99 characters)

cboulay commented:
OK this is ready for review. If you want a sample data file then that will have to wait until Friday when I'm in front of the hardware needed to produce the 3.0-ptp variant file spec.

samuelgarcia commented:
Hi Chadwick.
An enormous thanks for figthing with this new version of blackrock.
I will have a look soon. Lets wait friday to get the the test file that corresponf to this.
Can you try to make a very very short one ?

cboulay commented:
I found and fixed a bug in my previous implementation thanks to unit tests I just added.

Here's the test file. If you can provide me instructions on how to created a pull request on GIN then I can go ahead and do that, but I'm also totally satisfied having someone else do it if you're already setup for that.

[blackrock_3_0_ptp.zip](https://github.com/NeuralEnsemble/python-neo/files/13196454/blackrock_3_0_ptp.zip)

samuelgarcia commented:
I push tests files here https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/pulls/121/files

cboulay commented:
OK I've rebased and fixed the conflicts along the way. Let's hope the tests pass. BTW, I don't mind if you squash the commits when merging... there was some back and forth on quotes that become obsolete when you adopted PEP8.

cboulay commented:
Actually I went ahead and squashed into 1 commit.

The unit test isn't running because it can't find the file. I guess I'm not providing its path correctly. Can anyone help me out with that part?

zm711 commented on neo/test/rawiotest/test_blackrockrawio.py:
```diff
@@ -184,6 +185,32 @@ def test_compare_blackrockio_with_matlabloader_v21(self):
                         assert_equal(python_digievents, matlab_digievents)
 
+    def test_blackrockrawio_ptp_timestamps(self):
+        dirname = self.get_local_path("blackrock/blackrock_3_0_ptp/20231027-125608-001")
```
```suggestion
        dirname = self.get_local_path("blackrock/blackrock_3_0_ptp")
```

It looks to me like Sam's PR of your data removed this part of the file so the folder structure is just without the ending part.

zm711 reviewed (commented):
@cboulay see my comment and see if it looks reasonable.

See test structure here:
https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/blackrock/blackrock_3_0_ptp

cboulay commented:
I'm pretty lost trying to figure out what's going on the with file loading. I'm happy to yield to someone else.

zm711 commented on neo/test/rawiotest/test_blackrockrawio.py:
```diff
@@ -184,6 +185,32 @@ def test_compare_blackrockio_with_matlabloader_v21(self):
                         assert_equal(python_digievents, matlab_digievents)
 
+    def test_blackrockrawio_ptp_timestamps(self):
+        dirname = self.get_local_path("blackrock/blackrock_3_0_ptp/20231027-125608")
```
```suggestion
        dirname = self.get_local_path("blackrock/blackrock_3_0_ptp")
```

zm711 replied:
I'll push this change myself and see if it fixes it.

zm711 replied:
Actually do you have let maintainers push turned on? I can't edit this file myself.

zm711 commented:
@cboulay,

We are planning on including this in the next release which will go out in the next few days so if you have the time to either implement my commit fix or give us access we will include otherwise we will plan on this for the next release.

cboulay commented:
I don't see the option to allow edits from maintainers. I found the [GH docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) that explains it but the option isn't on this page.

zm711 commented:
Normally there is a little check box on the right side of the bag just below all the info. When I first looked at your PR you had it turned on but it got switched off at some point. Either way we will see if tests pass now!

zm711 commented:
Seems like it failed. Could you doublecheck the gin link I gave you and make sure it has all the files necessary for blackrock?

cboulay commented:
Ah, it's because I made the PR from the lab org. [sigh](https://github.com/orgs/community/discussions/5634)

zm711 commented on neo/test/rawiotest/test_blackrockrawio.py:
```diff
@@ -28,6 +28,7 @@ class TestBlackrockRawIO(
         "blackrock/FileSpec2.3001",
         "blackrock/blackrock_2_1/l101210-001",
         "blackrock/blackrock_3_0/file_spec_3_0",
+        "blackrock/blackrock_3_0_ptp/20231027-125608",
```
I think this one also needed to be fixed!

zm711 commented:
So I thin there are two paths that need to be fixed. One at the top and one at the bottom. Since I can't push stuff here do you want a PR on your branch with the two file path fixes I think we need and we can test?

zm711 commented:
That's cool. One last check:

could you try this as the filename? Previously you didn't have the -001 right? so maybe it needs the full file name for the .nev file? 

20231027-125608-001

zm711 commented:
Perfect that worked! thanks for bearing with me! The dirname messed with me! We should really call that filename in the tests for this IO (not your fault, something we've been discussing #1455 recently).

cboulay commented:
Seems to be working. Thanks for your help!

zm711 commented:
@samuelgarcia tests are passing here, but you also said you would look over the code. Did you want to read this over first before we merge?

samuelgarcia commented:
Thanks a lot for this new add Chadwick.
And thanks a lot to wait for me.
this looks good to me.
